specs/experimental/fault-proof/stage-one/fault-dispute-game.md

```diff
@@ -314,38 +314,49 @@ Large preimage proposals allow for submitters to stream in a large preimage over
 commitments to the intermediate state of the `keccak256` function after absorbing/permuting the $1088$ bit block.
 This data is progressively merkleized on-chain as it is streamed in, with each leaf constructed as follows:
 
-```solidity
-/// @notice Returns a leaf hash to add to a preimage proposal merkle tree.
-/// @param input A single 136 byte chunk of the input.
-/// @param blockIndex The index of the block that `input` corresponds to in the full preimage's absorbtion.
-/// @param stateCommitment The hash of the full 5x5 state matrix *after* absorbing and permuting `input`.
-function hashLeaf(
-    bytes memory input,
-    uint256 blockIndex,
-    bytes32 stateCommitment
-) internal view returns (bytes32 leaf) {
-    require(input.length == 136, "input must be exactly the size of the keccak256 rate");
-
-    leaf = keccak256(abi.encodePacked(input, blockIndex, stateCommitment));
-}
-```
+**Large Preimage Proposals**
+
+Large preimage proposals allow for submitters to stream in a large preimage over multiple transactions, along-side
+commitments to the intermediate state of the `keccak256` function after absorbing/permuting the $1088$ bit block.
+This data is available to off-chain challenge agents in blobs, and each blob contains up to $738$ intermediate SHA3
+commitments. The data is laid out as follows:
```

Comment:
I'm a little unclear exactly how this layout maps to the field elements in blobs. Ideally we'd use the same encoding as for derivation since we already have code to support that (https://specs.optimism.io/protocol/derivation.html#blob-encoding) and then once the data is converted from field elements to its continuous byte-string I assume we then read 176 byte chunks of it and extract the keccak block and post-state commitment as per the diagram.

Reply:
Also, to reassemble the data correctly, the challenger needs to know the sender and uuid the data is related to.  Previously this could be retrieved from the tx signer and uuid in the call data but it will now have to be either recorded by the contract or included in the blob data (and then verified by the contract).

The challenger also needs to find the _beacon_ block the blob is from, either by beacon block root or slot.  The execution block timestamp can be converted to beacon slot (so could load the execution block header by block number to get the timestamp).

The challenger will also need a way to determine which transaction in the block actually posted the content - we need to avoid the challenger being fooled by a tx with valid preimage data that didn't ever call the actual preimage oracle.

Reply:
Ah, block number/hash/timestamp and versioned kzg commitment the data was read from would be enough to identify blobs that were actually read - we'd just have to load all blobs from the block and we can then filter by commitment.

Still need a way to identify the sender and uuid the blob data was for though.

Reply:
In looking at it further I came around to the idea of just adjusting `mapping(address => mapping(uint256 => uint64[])) public proposalBlocks;` which currently maps `sender -> uuid -> block_number` to map to a `block_number, blob_hash` tuple.  We can find all the entries in that mapping by iterating the `proposals` array so it should be a pretty minimal change and gives us all the required info.